composer/loss/loss.py

```diff
@@ -0,0 +1,105 @@
+from __future__ import annotations
+
+from typing import Optional
+
+from torch import Tensor
+from torch.nn import functional as F
+
+__all__ = ["soft_cross_entropy"]
+
+
+def _infer_target_type(input: Tensor, targets: Tensor) -> str:
+    """Infers whether the target is in indices format or one_hot format.
+
+    Example indices format: [1, 4, 7] Example one_hot format [[0, 1, 0], [1, 0, 0], ...]
+    """
+    if input.shape == targets.shape:
+        return 'one_hot'
+    elif input.ndim == targets.ndim + 1:
+        return 'indices'
+    else:
+        raise RuntimeError(f'Unable to infer indices or one_hot. Targets has shape {targets.shape}'
+                           f' and the inputs to cross entropy has shape {input.shape}. For one_hot, '
+                           'expect targets.shape == inputs.shape. For indices, expect '
+                           'inputs.ndim == targets.ndim + 1')
+
+
+def ensure_targets_one_hot(input: Tensor, targets: Tensor) -> Tensor:
+    if _infer_target_type(input, targets) == 'indices':
+        targets = F.one_hot(targets, num_classes=input.shape[1])
+    return targets
+
+
+def _check_for_index_targets(targets: Tensor) -> bool:
+    """Checks if a given set of targets are indices by looking at the type."""
+    index_types = ['torch.LongTensor', 'torch.cuda.LongTensor']
```

Comment:
Can we compare against the dtype? Not a huge fan of string comparisons for type checks.

```suggestion
    index_types = [torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64]
```